dali/kernels/signal/fft/fft_cpu_impl_ffts.cc

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/kernels/signal/fft/fft_cpu_impl_ffts.h"
+#include <ffts/ffts.h>
+#include <cmath>
+#include <complex>
+#include <utility>
+#include <vector>
+#include "dali/core/common.h"
+#include "dali/core/convert.h"
+#include "dali/core/error_handling.h"
+#include "dali/core/format.h"
+#include "dali/core/util.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/signal/fft/fft_cpu_impl_utils.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+namespace fft {
+namespace impl {
+
+template <typename OutputType, typename InputType, int Dims>
+KernelRequirements Fft1DImplFfts<OutputType, InputType, Dims>::Setup(
+    KernelContext &context,
+    const InTensorCPU<InputType, Dims> &in,
+    const FftArgs &args) {
+  constexpr bool is_complex_out = std::is_same<OutputType, std::complex<float>>::value;
+  constexpr bool is_real_out = std::is_same<OutputType, float>::value;
+  DALI_ENFORCE((is_complex_out && args.spectrum_type == FFT_SPECTRUM_COMPLEX)
+            || (is_real_out && args.spectrum_type != FFT_SPECTRUM_COMPLEX),
+    "Output type should be complex<float> or float depending on the requested spectrum type");
+
+  transform_axis_ = args.transform_axis >= 0 ? args.transform_axis : Dims-1;
+  DALI_ENFORCE(transform_axis_ >= 0 && transform_axis_ < Dims,
+    make_string("Transform axis ", transform_axis_, " is out of bounds [0, ", Dims, ")"));
+
+  const auto n = in.shape[transform_axis_];
+  nfft_ = args.nfft > 0 ? args.nfft : n;
+  DALI_ENFORCE(nfft_ >= n, "NFFT is too small");
+
+  KernelRequirements req;
+  auto out_shape = in.shape;
+
+  ScratchpadEstimator se;
+  se.add<float>(AllocType::Host, size_in_buf(nfft_),  32);
+  se.add<float>(AllocType::Host, size_out_buf(nfft_), 32);
+  req.scratch_sizes = se.sizes;
+
+  if (args.spectrum_type == FFT_SPECTRUM_COMPLEX) {
+    out_shape[transform_axis_] = nfft_;
+  } else {
+    out_shape[transform_axis_] = nfft_/2 + 1;
+  }
+  req.output_shapes = {TensorListShape<DynamicDimensions>({out_shape})};
+
+  if (!plan_ || plan_n_ != nfft_) {
```

Comment:
What is `plan_n_` for?
I see it is set only in L76?

Reply:
Used in line 143. It is the ffts data structure necessary to compute an FFT of given size

Reply:
will remove